core/src/main/java/apoc/index/SchemaIndex.java

```diff
@@ -137,48 +144,40 @@ private void scanIndex(BlockingQueue<PropertyValueCount> queue, IndexDefinition
                 throw e;
             }
             if (isFullText(indexDefinition)) {
+                // similar to db.index.fulltext.queryNodes procedure
                 read.nodeIndexSeek(ktx.queryContext(), indexSession, cursor, IndexQueryConstraints.unconstrained(), PropertyIndexQuery.fulltextSearch("*"));
             } else {
                 read.nodeIndexScan(indexSession, cursor, IndexQueryConstraints.unorderedValues());
             }
-
-            Value previousValue = null;
-            long count = 0;
-            Map<Object, Integer> fullTextMap = new HashMap<>();
+            
+            Map<String, Map<Object, Integer>> valueCountMap = new HashMap<>();
+            
+            final Iterable<Label> labels = lblName.isEmpty() 
+                    ? indexDefinition.getLabels() 
+                    : Collections.singleton(Label.label(lblName));
+            
             while (cursor.next()) {
-                if (isFullText(indexDefinition)) {
-                    long nodeReference = cursor.nodeReference();
-                    FulltextProcedures.NodeOutput nodeOutput = FulltextProcedures.NodeOutput.forExistingEntityOrNull(tx, nodeReference, cursor.score());
-                    final Node node = nodeOutput.node;
-                    indexDefinition.getPropertyKeys()
-                            .forEach(propKey -> {
-                                final Object property = node.getProperty(propKey, null);
-                                if (property != null) {
-                                    fullTextMap.merge(property, 1, Integer::sum);
-                                }
-                            });
-                } else {
-                    for (int i = 0; i < cursor.numberOfProperties(); i++) {
-                        Value v = cursor.propertyValue(i);
-                        if (Objects.equals(v, previousValue)) { //  nullsafe equals
-                            count++;
-                        } else {
-                            if (previousValue != null) {
-                                putIntoQueue(queue, indexDefinition, key, previousValue.asObject(), count);
-                            }
-                            previousValue = v;
-                            count = 1;
-                        }
+                final Node node = threadTx.getNodeById( cursor.nodeReference() );
+
+                labels.forEach(label -> {
+
+                    final boolean hasLabel = node.hasLabel(label);
+                    
+                    final Object property = node.getProperty(key, null);
```

Comment:
Don't know if it make a noticable difference, but I guess for a slightly better performance, you could move this line before the labels.foreach loop.

Reply:
Changed